include/boost/geometry/algorithms/detail/overlay/traversal.hpp

```diff
@@ -606,13 +602,25 @@ public :
             return 0;
         }
 
-        return ranked_point.turn_index == start_turn_index
-                && ranked_point.operation_index == start_op_index ? 3
-            : ranked_point.turn_index == start_turn_index ? 2
-            : 1
+        bool const to_start = ranked_point.turn_index == start_turn_index;
+        bool const to_start_index = ranked_point.operation_index == start_op_index;
+
+        bool const next_in_same_cluster
+                = cluster_indices.count(op.enriched.get_next_turn_index()) > 0;
```

Comment:
This is the fix. In a cluster, it should prefer a way out of the cluster.

Reply:
`cluster_indices` is of type `std::set<signed_size_type>` which means that `count` returns either `0` or `1`. Is that what was desired or did you mean to use `std::multi_set` or `std::map<signed_size_type, size_t>` to count the indexes or maybe even `std::unordered_map<signed_size_type, size_t>` ?

Reply:
Indeed, `count` returns `1` if it is in the set, and that is what I'm using (I like it more than checking `find` and `end`)
So it should not be a map or a multiset

Reply:
Ok, you probably want to check whether or not an id was found, and how many times is not important. Correct?

Then I'd only suggest to think about `unordered_set<>` because the complexity of the lookup is O(1) on average, instead of O(logN). And AFAIU you only care to find if the id exists and sorting is not important. On the other hand it has higher constant cost than `set<>` and will become faster above some number of ids. So it depends on how many clusters do you expect to deal with.

Reply:
Sure, correct.

The number of clusters is usually very low. It’s where turns are colocated and that does not happen often in nature. (In our testsets it happens more because they are constructed for exactly that reason, these cases are harder.)

So just `set` is probably fine then. Thanks for your view.